### DISCUSSION TITLE
Mojo::Phantom to automate a web app

### DISCUSSION THREAD

**Discussion by hussam-qasem** [mojolicious/mojo#2069] [Q&A]
Is it possible to use @jberger [`Mojo::Phantom`](https://metacpan.org/pod/Mojo::Phantom) to drive a [`PhantomJS`](https://phantomjs.org/download.html) browser (and automate a web application)?

I would like to expose a [Mojolicious] API that then fetches data from a backend JavaScript web application:
```
                                Server
                               ┌────────────────────────────────────────────────┐
                               │                                                │
 Client                        │                          Mojo::Phantom         │       Legacy WebApp
┌──────────────┐               │                       ┌──────────────────────┐ │      ┌──────────────┐
│              │               │                       │                      │ │      │              │
│ Postman      │ POST Request  │                       │  Login (JS)       ───┼─┼─────►│ JavaScript   │
│ for example  ├──────────────►│                       │                      │ │      │ application  │
│              │               │                       │  Expand Menu (JS) ───┼─┼─────►│              │
│              │               │  Mojolicious App      │                      │ │      │              │
│              │               │                       │  SelectElement(JS)───┼─┼─────►│              │
│              │               │                       │                      │ │      │              │
│              │ JSON Response │                       │  Send Command     ───┼─┼─────►│              │
│              │◄──────────────┤                       │                      │ │      │              │
│              │               │                       │  Receive Result   ◄──┼─┼──────┤              │
│              │               │                       │                      │ │      │              │
└──────────────┘               │                       └──────────────────────┘ │      └──────────────┘
                               │                                                │
                               └────────────────────────────────────────────────┘
```

Few questions that come to mind:
* Can `Mojo::Phantom` help me save the resulting web page as an image (for debugging)?
* Is it possible to interact with PhantomJS (e.g., change the screen resolution)?
* Can I run execute JavaScript code and store the JS output in a Perl data structure? (e.g., the result of [`click_get()`](https://github.com/jberger/TestingWithPhantom/blob/0815d4db5dd17f298c333478421537ef7e045cff/ex/click.t#L13-L16) in [`click.t`](https://github.com/jberger/TestingWithPhantom/blob/master/ex/click.t) is not stored)
* Does `Mojo::Phantom` support "click" actions? (e.g. to handle `onclick` or `onsubmit` events)

Any advice and recommendation is much appreciated. Thank you!

**Comment by jberger** [comment]
In theory that would work, but PhantomJS is dead and thus so is Mojo::Phantom. There was some work around using headless chrome but their documentation was so bad that that was abandoned too.

**Reply by hussam-qasem** [reply]
Thank you @jberger for your prompt response. Do any of these functionalities exist today? I'll happily use `Mojo::Phantom` as-is. 

From my little research, `Mojo::Phantom` appears to be the lightest weight of all: [`WWW::Mechanize::PhantomJS`](https://metacpan.org/pod/WWW::Mechanize::PhantomJS), [`Selenium::PhantomJS`](https://metacpan.org/pod/Selenium::PhantomJS), etc.

Your input is highly appreciated. Thank you for your time.

**Comment by hussam-qasem** [comment]
I saw this recent article about [headless Chrome](https://developer.chrome.com/articles/new-headless/), so I'll give it a go and report back.

(I'm just falling in love with the [`Mojo::UserAgent`](https://docs.mojolicious.org/Mojo/UserAgent) and everything Mojo!)

**Comment by jberger** [comment]
Perhaps someday we can have a [Mojo::Chrome](https://github.com/jberger/Mojo-Chrome) though I see that that still uses Delays rather than async/await, so that shows how old it is.
